src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -127,16 +131,34 @@ void BatchedThreadedNnet3CudaOnlinePipeline::AllocateAndInitializeData(
 
 void BatchedThreadedNnet3CudaOnlinePipeline::SetLatticeCallback(
     CorrelationID corr_id, const LatticeCallback &callback) {
-  std::lock_guard<std::mutex> lk(map_callbacks_m_);
-  lattice_callbacks_.insert({corr_id, callback});
+  SegmentedResultsCallback segmented_callback =
+      [=](SegmentedLatticeCallbackParams params) {
+        if (params.results.empty()) {
+          KALDI_WARN << "Empty result for callback";
+          return;
+        }
+        CompactLattice &clat = params.results[0].GetLatticeResult();
+        callback(clat);
```

Comment:
SetLatticeCallback doesn't assume the caller will keep the callback in scope - in other Set*Callback functions, the copy of the callback happens when inserting in the map. So we'd need to make at least one copy of the argument anyway.

It is true that if we use SetLatticeCallback with a LatticeCallback (and not with a SegmentedResultsCallback) we perform two copies : one with the lambda capture, then one with the map insert. 

We could use a 

`SetLatticeCallback(CorrelationID corr_id, SegmentedResultsCallback &&callback, const int result_type) `

with a std::forward so that we could pass the SegmentedResultsCallback as a rvalue. That would avoid the second copy.

Reply:
@hugovbraun, while playing with the idea of rvalue reference at godbolt.com (I suspect we can avoid _both_ copies), I noticed a type mismatch between the type of lambda here
```c++
  SegmentedResultsCallback segmented_callback =
      [=](SegmentedLatticeCallbackParams params) {
// . . .
      };
```
and the definition of `SegmentedResultsCallback` (cudadecoder/cuda-pipeline-common.h:132)
```c++
typedef std::function<void(SegmentedLatticeCallbackParams &params)>
    SegmentedResultsCallback;
```
Looks like a call to the callback will always result in constructing a temporary copy of `SegmentedLatticeCallbackParams`, which has a whole vector of `CudaPipelineResult` with `CompactLattice` and other heavyweight stuff in it, and this is probably unintended. To verify, I declared the copy constructor of `CudaPipelineResult` as deleted, and it indeed failed to compile at line 134, where a `SegmentedResultsCallback` is constructed from the lambda.

Also, if the callback is really supposed to modify its argument, `SegmentedLatticeCallbackParams`, please change it to a pointer, otherwise to a const reference. We never use mutable references (except std streams).

It also may make sense to use `KALDI_DISALLOW_COPY_AND_ASSIGN` in the declarations of  both `CudaPipelineResult` and `SegmentedLatticeCallbackParams`, since they are probably not supposed to be copied; they are too heavyweight.

Reply:
An rvalue sample code for both `LatticeCallback` and `SegmentedResultsCallback`: https://godbolt.org/z/qz1r4M5hs. Doesn't look pretty, but eliminates copying of any `function` instance.

In any case, please don't sit too much on this. It's OK to step into the ugly side of C++ by me, as this is still WIP. Let's make it as efficient as possible ATM, as it may not be in the the hot path but is in quite a warm path indeed; then I'll think about a possible refactoring to make it more readable w/o perf loss. I think that the lambda with a lambda closure in its closure is not the most readable structure, and that's the key to untangling the ensuing ugliness. :)

 I'm using the online pipeline in production, so I'll be perfing it anyway.

Reply:
Great points, thank you, I'll change that

Reply:
It's done. We could look at using universal reference setter at some point, even if that'd require some templating in the header. 
For SegmentedLatticeCallbackParams, I guess we can use a rvalue ref for that kind of use. In the sense "I'm done using this object, do whatever you want with it". The idea is the user can modify the lattice or else if needed, without copying it first. That would work with a C pointer, but that doesn't include the meaning "the caller is done using this object, you can modify it if needed". What do you think ?